zkml/src/layers/transformer/mod.rs

```diff
@@ -174,61 +221,86 @@ mod test {
             input: &Tensor<f32>,
             gpt2_output: Option<&GPT2LayerOutput>,
         ) -> anyhow::Result<Tensor<f32>> {
-            assert_eq!(input.get_shape().len(), 2);
-            let seq_len = input.get_shape()[0];
+            ensure!(input.get_shape().len() == 2);
+
             let normed = self
                 .layernorm
                 .evaluate::<GoldilocksExt2>(&vec![input], vec![])?;
+
             if let Some(gpt2_output) = gpt2_output {
-                assert!(gpt2_output.is_layernorm_close(normed.outputs()));
+                ensure!(gpt2_output.is_layernorm_close(normed.outputs()));
             }
             let qkv = self
                 .qkv
                 .evaluate::<GoldilocksExt2>(&normed.outputs(), vec![])?;
+
             if let Some(gpt2_output) = gpt2_output {
-                assert!(gpt2_output.is_qkv_close(qkv.outputs()));
+                ensure!(gpt2_output.is_qkv_close(qkv.outputs()));
             }
             let mha = self
                 .mha
                 .evaluate::<GoldilocksExt2>(&qkv.outputs(), vec![])?;
+            // first, we need to apply causal mask before softmax
+            // let causal_mask = causal_mask(self.num_heads, seq_len, seq_len);
+            // let masked_attention_scores = mha.outputs()[0].add(&causal_mask);
```

Comment:
Leftover? Now causal mask is inside MHA layer, no?

Reply:
Yes